File: java/src/main/dsa/arrays/ValidAnagram.java
package dsa.arrays;

import java.util.HashMap;
import java.util.Map;

// https://neetcode.io/problems/is-anagram/question

public class ValidAnagram {

    public boolean isAnagram(String s, String t) {
        Map<Character, Integer> map = new HashMap<>();
        
        for (char ch : s.toCharArray()) {
            int val = map.getOrDefault(ch, 0) + 1;
            map.put(ch, val);
        
        }

        for (char ch : t.toCharArray()) {
            int val = map.getOrDefault(ch, 0) - 1;
            if (val < 0) {
                return false;
            } else if (val == 0) {
                map.remove(ch);
            } else {
                 map.put(ch, val);
            }
        }   

        return map.size() == 0;
    }

    public static void main(String[] args) {
        
        ValidAnagram validAnagram = new ValidAnagram();
        System.out.println(validAnagram.isAnagram("anagram", "nagaram")); // true
        System.out.println(validAnagram.isAnagram("rat", "car")); // false
    }
    
    
}

File: python/app/dsa/arrays/ValidAnagram.py

# Question : https://neetcode.io/problems/is-anagram/question

class Solution:
    
    def isAnagram(self,  s:str, t: str) -> bool:
        m = {}

        for s1 in s:
            val = m.get(s1, 0) + 1
            m[s1] = val

        for s2 in t:
            val = m.get(s2, 0) - 1
            if val < 0 :
                return False
            elif val == 0 :
                del m[s2]
            else :
                m[s2] = val

        return len(m) == 0   

if __name__ == '__main__' :
    sol = Solution()
    print(sol.isAnagram("racecar", "carrace"))
    print(sol.isAnagram("jar", "jam"))

             
